File: ebook/src/main/java/br/com/ebook/foobnix/ext/CacheZipUtils.java
package br.com.ebook.foobnix.ext;

import android.content.Context;

import androidx.core.util.Pair;

import org.ebookdroid.BookType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Enumeration;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipOutputStream;

import br.com.ebook.Config;

public class CacheZipUtils {
    private static final Logger LOGGER = LoggerFactory.getLogger(CacheZipUtils.class);
    private static final int BUFFER_SIZE = 16 * 1024;

    public enum CacheDir {
        ZipApp("ZipApp"), //
        ZipService("ZipService"); //

        private final String type;

        private CacheDir(String type) {
            this.type = type;
        }

        public static File parent;

        public String getType() {
            return type;
        }

        public static void createCacheDirs() {
            for (CacheDir folder : values()) {
                File root = new File(parent, folder.getType());
                if (!root.exists()) {
                    root.mkdirs();
                }
            }
        }


        public void removeCacheContent() {
            try {
                removeFiles(getDir().listFiles());
            } catch (Exception e) {
                LOGGER.error("Error to remove cache: {}", e.getMessage(), e);
            }
        }

        public File getDir() {
            return new File(parent, type);
        }

    }

    public static File CACHE_UN_ZIP_DIR;
    public static File CACHE_BOOK_DIR;
    public static File CACHE_WEB;
    public static File ATTACHMENTS_CACHE_DIR;
    public static final Lock cacheLock = new ReentrantLock();

    public static void init(Context c, File dir) {
        CACHE_BOOK_DIR = new File(dir, "Book");
        CACHE_UN_ZIP_DIR = new File(dir, "UnZip");
        ATTACHMENTS_CACHE_DIR = new File(dir, "Attachments");
        CACHE_WEB = new File(dir, "Web");

        CacheZipUtils.createAllCacheDirs();
        CacheDir.createCacheDirs();
    }

    public static void createAllCacheDirs() {
        if (!CACHE_BOOK_DIR.exists()) {
            CACHE_BOOK_DIR.mkdirs();
        }
        if (!ATTACHMENTS_CACHE_DIR.exists()) {
            ATTACHMENTS_CACHE_DIR.mkdirs();
        }
        if (!CACHE_WEB.exists()) {
            CACHE_WEB.mkdirs();
        }
    }

    public static void removeFiles(File[] files) {
        try {
            if (files == null) {
                return;
            }
            for (File file : files) {
                if (file != null) {
                    file.delete();
                }
            }
        } catch (Exception e) {
            LOGGER.error("Error to remove files: {}", e.getMessage(), e);
        }
    }


    public static void removeFiles(File[] files, File exept) {
        try {
            if (files == null || exept == null) {
                return;
            }
            for (File file : files) {
                if (file != null && !file.getName().startsWith(exept.getName())) {
                    file.delete();
                }
            }
        } catch (Exception e) {
            LOGGER.error("Error to remove files: {}", e.getMessage(), e);
        }
    }

    public static Pair<Boolean, String> isSingleAndSupportEntryFile(File file) {
        try {
            return isSingleAndSupportEntry(file);
        } catch (Exception e) {
            return new Pair<>(false, "");
        }
    }

    public static Pair<Boolean, String> isSingleAndSupportEntry(File is) {
        if (is == null) {
            return new Pair<>(false, "");
        }
        String name = "";
        try {
            ZipFile zipFile = new ZipFile(is, StandardCharsets.UTF_8);
            boolean find = false;
            ZipEntry nextEntry = null;

            Enumeration<? extends ZipEntry> entries = zipFile.entries();
            while (entries.hasMoreElements() && (nextEntry = entries.nextElement()) != null) {
                name = nextEntry.getName();
                if (find) {
                    zipFile.close();
                    return new Pair<>(false, "");
                }
                find = true;
            }
            zipFile.close();
        } catch (Exception e) {
            LOGGER.error("Error to valide is suport a single entry: {}", e.getMessage(), e);
        }
        return new Pair<>(BookType.isSupportedExtByPath(name), name);
    }

    public static class UnZipRes {
        public String originalPath;
        public String unZipPath;
        public String entryName;

        public UnZipRes(String originalPath, String unZipPath, String entryName) {
            this.originalPath = originalPath;
            this.unZipPath = unZipPath;
            this.entryName = entryName;
        }
    }

    public static UnZipRes extracIfNeed(String path, CacheDir folder) {
        if (!path.endsWith(".zip"))
            return new UnZipRes(path, path, null);

        folder.removeCacheContent();

        try {
            File file = new File(path);
            if (!isSingleAndSupportEntry(file).first)
                return new UnZipRes(path, path, null);

            ZipFile zipFile = new ZipFile(file, StandardCharsets.UTF_8);

            ZipEntry nextEntry = null;
            Enumeration<? extends ZipEntry> entries = zipFile.entries();
            while (entries.hasMoreElements() && (nextEntry = entries.nextElement()) != null) {
                if (BookType.isSupportedExtByPath(nextEntry.getName())) {
                    File out = new File(folder.getDir(), nextEntry.getName());
                    BufferedOutputStream fileOutputStream = new BufferedOutputStream(new FileOutputStream(out));
                    writeToStream(zipFile.getInputStream(nextEntry), fileOutputStream);
                    if (Config.SHOW_LOG)
                        LOGGER.info("Unpack archive: {}", file.getPath());

                    zipFile.close();
                    return new UnZipRes(path, file.getPath(), nextEntry.getName());
                }
            }
            zipFile.close();
        } catch (Exception e) {
            LOGGER.error("Error to extract: {}", e.getMessage(), e);
        }
        return new UnZipRes(path, path, null);
    }

    public static void writeToStream(InputStream zipInputStream, OutputStream out) throws IOException {

        byte[] bytesIn = new byte[BUFFER_SIZE];
        int read = 0;
        while ((read = zipInputStream.read(bytesIn)) != -1) {
            out.write(bytesIn, 0, read);
        }
        out.close();
    }

    static public void zipFolder(String srcFolder, String destZipFile) throws Exception {
        ZipOutputStream zip = null;
        FileOutputStream fileWriter = null;

        fileWriter = new FileOutputStream(destZipFile);
        zip = new ZipOutputStream(fileWriter);
        zip.setLevel(0);

        addFolderToZip("", srcFolder, zip);
        zip.flush();
        zip.close();
    }

    static private void addFileToZip(String path, String srcFile, ZipOutputStream zip) throws Exception {

        File folder = new File(srcFile);
        if (folder.isDirectory()) {
            addFolderToZip(path, srcFile, zip);
        } else {
            byte[] buf = new byte[BUFFER_SIZE];
            int len;
            FileInputStream in = new FileInputStream(srcFile);
            zip.putNextEntry(new ZipEntry(path + "/" + folder.getName()));
            while ((len = in.read(buf)) > 0) {
                zip.write(buf, 0, len);
            }
            in.close();
        }
    }

    static private void addFolderToZip(String path, String srcFolder, ZipOutputStream zip) throws Exception {
        File folder = new File(srcFolder);

        for (String fileName : folder.list()) {
            if (path.equals(""))
                addFileToZip(folder.getName(), srcFolder + "/" + fileName, zip);
            else
                addFileToZip(path + "/" + folder.getName(), srcFolder + "/" + fileName, zip);
        }
    }

    public static void deleteDir(File file) {
        File[] contents = file.listFiles();
        if (contents != null) {
            for (File f : contents) {
                deleteDir(f);
            }
        }
        file.delete();
    }

    public static void copyFile(File source, File dest) throws IOException {
        InputStream is = null;
        OutputStream os = null;
        try {
            is = new FileInputStream(source);
            os = new FileOutputStream(dest);
            byte[] buffer = new byte[BUFFER_SIZE];
            int length;
            while ((length = is.read(buffer)) > 0) {
                os.write(buffer, 0, length);
            }
        } finally {
            is.close();
            os.close();
        }
    }
}

File: ebook/src/main/java/br/com/ebook/foobnix/ext/CacheZipUtils.kt
package br.com.ebook.foobnix.ext

import android.content.Context
import androidx.core.util.Pair
import br.com.ebook.util.IOUtils
import br.com.ebook.util.IOUtils.copyTo
import org.ebookdroid.BookType
import org.slf4j.LoggerFactory
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import java.util.concurrent.locks.ReentrantLock
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipOutputStream

object CacheZipUtils {
    private val LOGGER = LoggerFactory.getLogger(CacheZipUtils::class.java)

    enum class CacheDir(val type: String) {
        ZipApp("ZipApp"),
        ZipService("ZipService");

        fun removeCacheContent() {
            try {
                removeFiles(getDir().listFiles())
            } catch (e: Exception) {
                LOGGER.error("Error to remove cache: {}", e.message, e)
            }
        }

        fun getDir(): File {
            val p = parent ?: throw IllegalStateException("Parent cache directory is not initialized")
            return File(p, type)
        }

        companion object {
            @JvmField
            var parent: File? = null

            @JvmStatic
            fun createCacheDirs() {
                val p = parent ?: return
                values().forEach { folder ->
                    val root = File(p, folder.type)
                    if (!root.exists()) {
                        root.mkdirs()
                    }
                }
            }
        }
    }

    @JvmField
    var CACHE_UN_ZIP_DIR: File? = null
    
    @JvmField
    var CACHE_BOOK_DIR: File? = null
    
    @JvmField
    var CACHE_WEB: File? = null
    
    @JvmField
    var ATTACHMENTS_CACHE_DIR: File? = null
    
    @JvmField
    val cacheLock = ReentrantLock()

    @JvmStatic
    fun init(c: Context, dir: File) {
        CACHE_BOOK_DIR = File(dir, "Book")
        CACHE_UN_ZIP_DIR = File(dir, "UnZip")
        ATTACHMENTS_CACHE_DIR = File(dir, "Attachments")
        CACHE_WEB = File(dir, "Web")

        CacheDir.parent = dir

        createAllCacheDirs()
        CacheDir.createCacheDirs()
    }

    @JvmStatic
    fun createAllCacheDirs() {
        CACHE_BOOK_DIR?.let { if (!it.exists()) it.mkdirs() }
        ATTACHMENTS_CACHE_DIR?.let { if (!it.exists()) it.mkdirs() }
        CACHE_WEB?.let { if (!it.exists()) it.mkdirs() }
    }

    @JvmStatic
    fun removeFiles(files: Array<File>?) {
        try {
            files?.forEach { file ->
                file.delete()
            }
        } catch (e: Exception) {
            LOGGER.error("Error to remove files: {}", e.message, e)
        }
    }

    @JvmStatic
    fun removeFiles(files: Array<File>?, exept: File?) {
        try {
            if (files == null || exept == null) return
            files.forEach { file ->
                if (!file.name.startsWith(exept.name)) {
                    file.delete()
                }
            }
        } catch (e: Exception) {
            LOGGER.error("Error to remove files: {}", e.message, e)
        }
    }

    @JvmStatic
    fun isSingleAndSupportEntryFile(file: File): Pair<Boolean, String> {
        return try {
            isSingleAndSupportEntry(file)
        } catch (e: Exception) {
            Pair(false, "")
        }
    }

    @JvmStatic
    fun isSingleAndSupportEntry(file: File?): Pair<Boolean, String> {
        if (file == null) return Pair(false, "")
        var name = ""
        try {
            ZipFile(file, StandardCharsets.UTF_8).use { zipFile ->
                var find = false
                val entries = zipFile.entries()
                while (entries.hasMoreElements()) {
                    val entry = entries.nextElement()
                    name = entry.name
                    if (find) {
                        return Pair(false, "")
                    }
                    find = true
                }
            }
        } catch (e: Exception) {
            LOGGER.error("Error to validate if zip has single entry: {}", e.message, e)
        }
        return Pair(BookType.isSupportedExtByPath(name), name)
    }

    class UnZipRes(
        @JvmField val originalPath: String, 
        @JvmField val unZipPath: String?, 
        @JvmField val entryName: String?
    )

    @JvmStatic
    fun extracIfNeed(path: String, folder: CacheDir): UnZipRes {
        if (!path.endsWith(".zip")) {
            return UnZipRes(path, path, null)
        }

        folder.removeCacheContent()

        try {
            val file = File(path)
            if (isSingleAndSupportEntry(file).first == false) {
                return UnZipRes(path, path, null)
            }

            ZipFile(file, StandardCharsets.UTF_8).use { zipFile ->
                val entries = zipFile.entries()
                while (entries.hasMoreElements()) {
                    val entry = entries.nextElement()
                    if (BookType.isSupportedExtByPath(entry.name)) {
                        val out = File(folder.getDir(), entry.name)
                        BufferedOutputStream(FileOutputStream(out)).use { fos ->
                            zipFile.getInputStream(entry).use { it.copyTo(fos) }
                        }
                        LOGGER.info("Unpack archive: {}", file.path)
                        return UnZipRes(path, file.path, entry.name)
                    }
                }
            }
        } catch (e: Exception) {
            LOGGER.error("Error to extract zip cache: {}", e.message, e)
        }
        return UnZipRes(path, path, null)
    }

    @JvmStatic
    fun writeToStream(zipInputStream: InputStream, out: OutputStream) {
        try {
            zipInputStream.use { input ->
                out.use { output ->
                    input.copyTo(output)
                }
            }
        } catch (e: Exception) {
            LOGGER.error("Error writing to stream: {}", e.message, e)
        }
    }

    @JvmStatic
    @Throws(Exception::class)
    fun zipFolder(srcFolder: String, destZipFile: String) {
        FileOutputStream(destZipFile).use { fileWriter ->
            ZipOutputStream(fileWriter).use { zip ->
                zip.setLevel(0)
                addFolderToZip("", srcFolder, zip)
                zip.flush()
            }
        }
    }

    @JvmStatic
    @Throws(Exception::class)
    private fun addFileToZip(path: String, srcFile: String, zip: ZipOutputStream) {
        val file = File(srcFile)
        if (file.isDirectory) {
            addFolderToZip(path, srcFile, zip)
        } else {
            val buf = ByteArray(IOUtils.BUFFER_SIZE)
            var len: Int
            FileInputStream(srcFile).use { fis ->
                zip.putNextEntry(ZipEntry(path + "/" + file.name))
                while (fis.read(buf).also { len = it } > 0) {
                    zip.write(buf, 0, len)
                }
            }
        }
    }

    @JvmStatic
    @Throws(Exception::class)
    private fun addFolderToZip(path: String, srcFolder: String, zip: ZipOutputStream) {
        val folder = File(srcFolder)
        folder.list()?.forEach { fileName ->
            val nextPath = if (path.isEmpty()) folder.name else "$path/${folder.name}"
            addFileToZip(nextPath, "$srcFolder/$fileName", zip)
        }
    }

    @JvmStatic
    fun deleteDir(file: File) {
        file.listFiles()?.forEach { f ->
            deleteDir(f)
        }
        file.delete()
    }

    @JvmStatic
    @Throws(IOException::class)
    fun copyFile(source: File, dest: File) {
        IOUtils.copyFile(source, dest)
    }
}
